docs/src/vjp_laws.jl

```diff
@@ -0,0 +1,377 @@
+#       glacier_idx,
+#   )
+#   ```
+#   - Intent and behavior:
+#     - This is an internal routine. It is NOT intended to be called by users directly.
+#     - It is invoked when ODINN must fall back to the AD backend (with [DifferentiationInterface.jl](https://github.com/JuliaDiff/DifferentiationInterface.jl)) because the law did not supply explicit VJP functions (`f_VJP_input!`/`f_VJP_θ!` or because `p_VJP!` is set to `DIVJP()`).
+#     - Its job is to precompile and prepare the AD-based VJP code for a given law and to produce *preparation* objects that store [preparation](https://juliadiff.org/DifferentiationInterface.jl/DifferentiationInterface/stable/explanation/operators/#Preparation) results.
+#     - `prepare_vjp_law` is typically called just after the iceflow model / law objects have been instantiated — i.e., early in setup — so that preparations are ready before solving or adjoint runs.
+
+# - `precompute_law_VJP` (used before solving the adjoint PDE)
+
+#   We have overloads in the codebase:
+#   ```
+#   precompute_law_VJP(
+#     law::AbstractLaw,
+#     cache,
+#     vjpsPrepLaw,
+#     simulation,
+#     glacier_idx,
+#     t,
+#     θ
+#   )
+#   ```
+#   - Intent and behavior:
+#     - This function precomputes VJP-related artifacts *before* the adjoint iceflow PDE is solved for given time `t` and parameters `θ`.
+#     - It typically uses the `vjpsPrepLaw` (an `AbstractPrepVJP` instance produced earlier by `prepare_vjp_law`) together with the `cache` and `simulation` object. The produced results are cached in `cache` and are optionally consumed later by `law_VJP_input` / `law_VJP_θ` during the adjoint solve.
+
+# - Entry points used in the adjoint PDE
+
+#   These functions are the actual runtime entry points used when computing contributions of the laws to the gradient in the adjoint PDE:
+#   ```
+#   law_VJP_θ(law::AbstractLaw, cache, simulation, glacier_idx, t, θ)
+#   ```
+#   and
+#   ```
+#   law_VJP_input(law::AbstractLaw, cache, simulation, glacier_idx, t, θ)
+#   ```
+#   - Intent and behavior:
+#     - These are called during the adjoint solve to compute parameter and input VJPs for the law at time `t` and for parameters `θ`.
+#     - They can either compute the VJPs directly or use cached VJP information that has been already computed in the user-supplied `p_VJP!` VJP function. The `cache` allows storing useful information from the forward or from the precomputation step.
+#     - They therefore carry the runtime context (simulation, glacier index, time, θ) which is necessary for adjoint calculations.
+
+# ### Workflow
+# How the pieces compose in practice?
+# - If you, as a user, provide custom VJP functions (through `f_VJP_input!`/`f_VJP_θ!`, or through `p_VJP!`), ODINN will use them directly at adjoint time and will skip the AD fallback path. You can also provide your own precompute wrapper and cache to optimize expensive computations.
+# - If you do NOT provide VJP functions, ODINN runs the AD fallback:
```

Comment:
How robust is this? Can we say it can generalize to multiple laws?

Something I have not seen is where in the ODINN API we can change the AD backend of DI.